### ISSUE TITLE
Provide a way for plugins to know if they're being run for the first time

### ISSUE THREAD

**Issue by shepazon** [matryer/xbar#609]
It would be useful in many cases for plugins to know

* If they're being run for the first time after being installed (so they can do things like create folders, set up files, check dependencies, etc).
* If they're being run for the first time during the current session (so they can avoid triggering periodic effects before the user expects them, and to let the plugin know it should reset any state variables that it might be keeping on disk)

**Comment by matopeto** [comment]
1. you can create some hidden file to determine if you run for first time (non exist -> first time)
2. what you mean by "current session"?

**Comment by shepazon** [comment]
1. Yes, that's true about the first run ever part.
2. By "first time during the current session" I mean the first time BitBar runs each script after being started up.

**Comment by matryer** [comment]
@shepazon interesting idea. Can you explain a little bit about why this would be useful in your case?

**Comment by shepazon** [comment]
In my specific case, I'm creating a plugin that's part of a multi-script bundle of code to manage corporate network VPN connections. The BitBar plugin's purpose is to show the current up/down state of the VPN, provide statistics about the connection in grey items in the menu (connection duration, how long until you get automatically cut off, etc), and Connect/Disconnect options.

The "First time ever" scenario would be useful because I could use that to automatically install all the scripts, binaries, packages, and LaunchAgent plist files needed to support things. That would let the BitBar plugin basically handle it all, instead of needing a custom installer like I have now that installs the BitBar plugin, creates the directory the binaries go into, downloads needed packages, etc.

Having a "First time for this version of the plugin" event/notification/signal would let me check to see if the tools are up to date, but only if necessary, saving some time during startup. It would also let me do things like update the format of configuration data as needed, and that kind of thing, if necessary (there's no need for it as things stand but I could see it in the future).

Both of those would also let me do things like handle any needed LaunchAgent setup that needs doing.

Having a "First time since BitBar started up" would let me do things like launch needed background tasks, and that kind of thing. I don't have to launch them on subsequent runs of my script, since those tasks are already running after the first pass.

Yeah, there are ways around some these, either by simulating these events using things like a marker file that I can create/delete to indicate first-run scenarios, but that can be pretty fragile. Having a proper mechanism to make my script aware of which scenario is the case would be very useful.

I can think of many other ways they would be handy, but you asked for my specific case. :)

**Comment by matopeto** [comment]
> Having a "First time since BitBar started up" would let me do things like launch needed background tasks, and that kind of thing. I don't have to launch them on subsequent runs of my script, since those tasks are already running after the first pass.

I think in this case it will not help you. I can quit bitbar and run it again and your services and tasks will be already running. Try determine if your tasks are running in different way. I think it is better and safer way. 

And first time for this version of the plugin. I think Bitbar doesn't known version of the plugin (yes it does when you provide it, but it is not mandatory) better is write down version to hidden file/user defaults, and compare when your script is called by bitbar and do migration if you need. 

I am not developer of bitbar only i do some plugin where i solve similar problems so this is my point of view.

**Comment by mlcampbe** [comment]
I am going to chime in here and see if anyone has a way to do this yet. In my case I've developed a plugin that detects if the macbook is docked or not based on the presence of an external monitor. If I find the monitor then I run 1 set of commands but if I don't find it I do a different set of commands.  But I don't need to run either set of commands but 1 time after switching from docked to undocked or vice verse.

I would like to initialize the icon with details and populate my status text file on the initial loading of the plugin. Then subsequent refreshes could detect if the current docking status is the same as the previous status there is no need to rebuild the menu nor execute any actions.

If I have undocked the macbook then the next script execution should detect that and run my commands and reset the status as "undocked". Subsequent refreshes after that do not need to do anything until the status changes to "docked" again.

Is there anyway to do this through variables perhaps?
